design/one-pager-easy-stacks.md

```diff
@@ -0,0 +1,294 @@
+    allocatedStorage: 20
+    engine: mysql
+    skipFinalSnapshotBeforeDeletion: true
+  providerRef:
+    name: aws-provider
+  reclaimPolicy: Delete
+```
+
+Other people in different namespaces can create their own instances of `EasyAWS`
+custom resource and have their own similar environment with different names.
+
+## Technical Implementation
+
+### Metadata of The Stack
+
+[Dependency list] has to be filled with each kind of the resource that the stack
+deploys, i.e. f the stack only deploys `RDSInstanceClass` resource, it should
+not require all AWS stack to be present.
+
+[UI annotations] should be present to make it easy for frontend software to process
+the stack.
+
+### Folder Structure
+
+All directories containing base YAML files should be in a top-level directory
+of the repository named `resources` so that it's easier for others to fork & change.
+
+Controller will scan this folder recursively to get all YAML files. So, the folder
+structure inside that `resources` directory won't matter to the controller but
+it's good to separate the folders cleanly so that it's easier to fork & change.
+
+There are mainly 2 types of resources that configuration stacks will usually
+need to deploy; resource classes and managed resources. Under `resources` folder
+we can have 2 separate folders for each of those resource types; `classes` and
+`services`. Under these folders, resources should be in directories recursively
+named after their group and kind. An example folder structure:
+
+```
+├── resources
+│   ├── classes
+│   │   ├── cache
+│   │   │   └── replicationgroup.yaml
+│   │   ├── compute
+│   │   │   └── ekscluster.yaml
+│   │   └── database
+│   │       ├── rdsinstancemysql.yaml
+│   │       └── rdsinstancepostgresql.yaml
+│   └── services
+│       ├── database
+│       │   └── dbsubnetgroup.yaml
+│       ├── identity
+│       │   ├── iamrole.yaml
+│       │   └── iamrolepolicyattachment.yaml
+│       └── network
+│           ├── internetgateway.yaml
+│           ├── routetable.yaml
+│           ├── securitygroup.yaml
+│           ├── subnet.yaml
+│           └── vpc.yaml
+```
+
+Example above is a suggestion for stacks with one tier of configuration.
+There could also be cases where one configuration stack has the YAML files for
+different tiers and a `spec` allows to deploy one of them, in that case it'd make
+sense for developer to have a folder for each tier under `resources` folder.
+
+### Reconciliation
+
+There are mainly 3 responsibility of the controller:
+* When all deployed resources are ready, `status.state` should be `ready`.
+  * For resource classes; if the resource exists in the api-server it's considered
+    to be ready.
+  * For managed resources; if the `Ready` condition status is `true`, then it's
+    considered to be ready.
+  * For other possible resources, the developer chooses what to look for as
+    readiness.
+* Apply; continuously reproduce the resources from YAMLs and apply it.
```

Comment:
Do you expect there will be one generic controller loading this YAML as `*unstructured.Unstructured`, or a different controller for each stack variant that knows the actual kinds it supports at the code level?

It doesn't seem like the controller will really care much about the shape of the resources with two exceptions - it will want to add metadata (e.g. set labels) when creating resources, and will need to know how each resource reflects it readiness.

Reply:
> Do you expect there will be one generic controller loading this YAML as *unstructured.Unstructured, or a different controller for each stack variant that knows the actual kinds it supports at the code level?

Became unrelated as we started to use `kustomize`. But in patchers, we do make the operations on `*unstructured.Unstructured` type.

> will need to know how each resource reflects it readiness.

No readiness anymore. If the reconciliation is completed with no errors, it means the resources are deployed and we tell that to user via `Synced` condition.